sentinelhub/data_utils.py

```diff
@@ -116,37 +122,47 @@ def statistical_to_dataframe(result_data: List[JsonDict], exclude_stats: Optiona
+    """Collect failed intervals of a single geometry from the (Batch) Statistical result
 
-    :param identifier: The identifier of the geometry.
-    :param response_data: An input representation of Statistical API response.
-    :return: The identifier of a geometry that has a response status of PARTIAL and the failed intervals.
+    :param response_data: An input representation of the (Batch) Statistical API response of a geometry.
+    :return: The failed intervals for a geometry.
     """
-    failed_intervals = []
-    for interval in response_data:
-        if "error" in interval:
-            failed_intervals.append((interval["interval"]["from"], interval["interval"]["to"]))
+    return [
+        (interval["interval"]["from"], interval["interval"]["to"]) for interval in response_data if "error" in interval
+    ]
 
-    return {"identifier": identifier, "failed_intervals": failed_intervals} if failed_intervals else None
+
+def _get_failed_batch_response(result_data: JsonDict) -> Union[str, List[Tuple[str, str]]]:
+    """Collect failed responses
+
+    :param result_data: An input representation of the (Batch) Statistical API result of a geometry.
+    :return: Failed responses and responses with failed intervals
+    """
+    if "error" in result_data or not result_data["response"]:
+        return _FULL_TIME_RANGE
+    return _get_failed_intervals(result_data["response"]["data"])
 
 
-def get_failed_statistical_requests(result_data: List[JsonDict]) -> List[Dict[str, Union[str, List[Tuple[str, str]]]]]:
+def get_failed_statistical_requests(result_data: List[JsonDict]) -> List[JsonDict]:
     """Collect failed requests of (Batch) Statistical Results
 
     :param result_data: An input representation of (Batch) Statistical API result.
     :return: Failed requests of (Batch) Statistical Results.
     """
-    failed_requests = []
-    for result in result_data:
-        identifier, response = result["identifier"], result["response"]
-        if not response:
-            failed_requests.append({"identifier": identifier})
-        else:
-            failed_intervals = _get_failed_intervals(identifier, response["data"])
-            if failed_intervals:
-                failed_requests.append(failed_intervals)
-
-    return failed_requests
+    if _is_batch_stat(result_data[0]):
+        batch_failed_response = ((result["identifier"], _get_failed_batch_response(result)) for result in result_data)
+        return [
+            {"identifier": ident, "failed_intervals": intervals}
+            for ident, intervals in batch_failed_response
+            if intervals != []
+        ]
+
+    normal_failed_intervals = (
+        _FULL_TIME_RANGE if "error" in result else _get_failed_intervals(result["data"]) for result in result_data
+    )
+    return [
+        {"identifier": index, "failed_intervals": intervals}
+        for index, intervals in enumerate(normal_failed_intervals)
+        if intervals != []
+    ]
```

Comment:
Excellent, now we can try and do this!
```suggestion
     if _is_batch_stat(result_data[0]):
        failed_responses = ((result["identifier"], _get_failed_batch_response(result)) for result in result_data)
    else:
        failed_responses = enumerate(
            _FULL_TIME_RANGE if "error" in result else _get_failed_intervals(result["data"]) for result in result_data
        )

    return [
        {"identifier": idx, "failed_intervals": intervals} for idx, intervals in failed_responses if intervals != []
    ]
```

Doublecheck this actually works :see_no_evil:

Reply:
It works but in this case mypy doesn't like the assignment of generator and enumerate to the same variable, which may lead to something looks bad like `failed_responses: Any = ...`

Is there a preferable way to solve this?

Reply:
Otherwise I see that `if else` in this case make the script much shorter and cleaner :)

Reply:
I would do this:
```python
    failed_responses: Iterable[Tuple[Any, Union[str, List[Tuple[str, str]]]]]

    if _is_batch_stat(result_data[0]):
        failed_responses = ((result["identifier"], _get_failed_batch_response(result)) for result in result_data)
    else:
        failed_responses = enumerate(
            _FULL_TIME_RANGE if "error" in result else _get_failed_intervals(result["data"]) for result in result_data
        )
    return [
        {"identifier": idx, "failed_intervals": intervals} for idx, intervals in failed_responses if intervals != []
    ]
```